### PR TITLE
feat: integration of AWS IAM Roles for ServiceAccounts (IRSA)

### PR DESCRIPTION
**Pull request by farshad-hobsons**
### Summary of Changes:

- BUG: Add functionality to use an existing service account - Same as https://github.com/godaddy/kubernetes-external-secrets/issues/170 &  https://github.com/godaddy/kubernetes-external-secrets/pull/169
TESTED with:
```
helm --tiller-namespace ops --namespace ops upgrade um-kubernetes-external-secret --set serviceAccount.name=um-farshad-test --set env.AWS_REGION=us-east-1 --set serviceAccount.create=false  --debug -i ./charts/kubernetes-external-secrets
```

BEFORE change incorrect serviceAccountName in rendered yaml:
```yaml
...
# Source: kubernetes-external-secrets/templates/deployment.yaml
...
    spec:
      serviceAccountName: um-kubernetes-external-secret-kubernetes-external-secrets
```
AFTER change correct serviceAccountName in rendered yaml:
```yaml
...
# Source: kubernetes-external-secrets/templates/deployment.yaml
...
    spec:
      serviceAccountName: um-farshad-test
```
Also TESTED with (for regression purposes):
```
helm --tiller-namespace ops --namespace ops upgrade um-kubernetes-external-secret  --set env.AWS_REGION=us-east-1  --debug -i ./charts/kubernetes-external-secrets
```

- BUG: Latest [1.5.0 version requires `GET` on namespace](https://github.com/godaddy/kubernetes-external-secrets/blob/58283fd6c6114ad3e19bf6f51da636ff63dfbc76/lib/poller.js#L107) which the current chart was not providing.

- FEATURE: Optionally Parameterize `securityContext`(needed for IRSA support [see reference issue](https://github.com/aws/amazon-eks-pod-identity-webhook/issues/8)) and `saAnnotations`(needed for [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/specify-service-account-role.html)) and `imagePullSecrets`(needed if someone wants to build their own image and deploy from a private repo)
TESTED with:
```
    helm --tiller-namespace ops --namespace ops upgrade um-kubernetes-external-secret --set env.AWS_REGION=us-east-1 --set securityContext.fsGroup=65534  --debug -i ./charts/kubernetes-external-secrets
    helm --tiller-namespace ops --namespace ops upgrade um-kubernetes-external-secret --set env.AWS_REGION=us-east-1 --set securityContext.fsGroup=65534 --set image.imagePullSecrets=TEST --debug -i ./charts/kubernetes-external-secrets
    helm --tiller-namespace ops --namespace ops upgrade um-kubernetes-external-secret --set env.AWS_REGION=us-east-1 --set securityContext.fsGroup=65534 --set image.imagePullSecrets=TEST --set saAnnotations."eks\.amazonaws\.com/role-arn"='arn:aws:iam::111111111111:role/ROLENAME'  --debug -i ./charts/kubernetes-external-secrets
```
- UPGRADE: aws-sdk-js from 2.433.0 to [v2.547.0](https://github.com/aws/aws-sdk-js/releases/tag/v2.547.0)

- IRSA CODE UPDATE:
  - Using the new IRSA feature or the existing assumeRole feature are mutually exclusive. Only one can work at a time based on if a roleArn is passed as an annotation to externalSecret object(AssumeRole functionality) or if IRSA environment variable is set to true. **Note that this mutual exclusiveness if irsa and assumeRole is due to the issue-2 pointed out in https://github.com/godaddy/kubernetes-external-secrets/issues/174. Addressing https://github.com/godaddy/kubernetes-external-secrets/issues/174 (issue-2) is a desired re-factor from architectural perspective**
  - Tested IRSA functionality with:
```
helm --tiller-namespace ops --namespace ops upgrade um-irsa-kubernetes-external-secret --set env.AWS_REGION=us-east-1 --set env.AWS_IRSA=true --set securityContext.fsGroup=65534 --set image.repository=<REDACTED> --set image.tag=feature-irsa-integration-7147b26 --set image.imagePullSecrets=<REDACTED> --set saAnnotations."eks\.amazonaws\.com/role-arn"='arn:aws:iam::111111111111:role/<REDACTED>'  --debug -i ./charts/kubernetes-external-secrets
```

### CODE REVIEW

**Comment by timcosta** [comment]
Hey @farshad-hobsons let me know if there's anything I can do to help move this along. Really looking forward to integrating this project into my cluster, but I need the IRSA support!

**Comment by imaffe** [comment]
@farshad-hobsons This is important for me as well, but we need someone has the write access to solve all the conflicts. Wondering when will be the next release, as IRSA is essential for this plugin.

File: charts/kubernetes-external-secrets/Chart.yaml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 apiVersion: v1
 name: kubernetes-external-secrets
-version: 1.0.1
-appVersion: 1.5.0
+version: 1.0.2
+appVersion: 1.6.0
```
Comment (Flydiverny) [inline charts/kubernetes-external-secrets/Chart.yaml]: No need to change these as the release script handles that.

File: charts/kubernetes-external-secrets/values.yaml [hunk 2]
```diff
@@ -29,18 +31,26 @@ serviceAccount:
...
 
+saAnnotations: {}
+  # eks.amazonaws.com/role-arn: arn:aws:iam::111111111111:role/ROLENAME
+
 replicaCount: 1
 
 image:
   repository: godaddy/kubernetes-external-secrets
-  tag: 1.5.0
+  tag: 1.6.0
```
Comment (Flydiverny) [inline charts/kubernetes-external-secrets/values.yaml]: release script will update this as well

File: charts/kubernetes-external-secrets/values.yaml [hunk 3]
```diff
@@ -29,18 +31,26 @@ serviceAccount:
...
+  imagePullSecrets: ""
 
 nameOverride: ""
 fullnameOverride: ""
 
 podAnnotations: {}
 
+securityContext: {}
+  # securityContext:
+  #  fsGroup: 65534
```
Comment (Flydiverny) [inline charts/kubernetes-external-secrets/values.yaml]: This is somewhat misleading please change to (similar to resources below)

```
securityContext: {}
  # fsGroup: 65534
```

File: lib/backends/kv-backend.js [hunk 4]
```diff
@@ -25,7 +25,11 @@ class KVBackend extends AbstractBackend {
    */
   _fetchSecretPropertyValues ({ externalData, roleArn }) {
     return Promise.all(externalData.map(async secretProperty => {
-      this._logger.info(`fetching secret property ${secretProperty.name} with role: ${roleArn}`)
+      if (typeof roleArn === 'undefined' && process.env.AWS_ROLE_ARN) {
```
Comment (Flydiverny) [inline lib/backends/kv-backend.js]: Better if we can pass in aws role arn instead of calling `process.env` here

File: lib/poller.js [hunk 5]
```diff
@@ -138,7 +138,7 @@ class Poller {
     // an empty annotation value allows access to all roles
     const re = new RegExp(namespace.metadata.annotations[annotationPermittedKey])
 
-    if (!re.test(role)) {
+    if (!re.test(role) && !re.test(process.env.AWS_ROLE_ARN.split('/').slice(-1)[0])) {
```
Comment (Flydiverny) [inline lib/poller.js]: See comments above

File: lib/backends/kv-backend.js (hunk 4 above)
Comment (Flydiverny) [reply lib/backends/kv-backend.js]: Maybe we can make this less intrusive by utilising `roleArn` here as well, ie if theres no `roleArn` set in the secretProperties we could default it to `AWS_ROLE_ARN` value. Perhaps we can sanitise the values in the Poller as its used in there as well.

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "kubernetes-external-secrets",
-  "version": "1.5.0",
+  "version": "1.6.0",
```
Comment (Flydiverny) [inline package.json]: no need to bump

File: config/aws-config.js [hunk 7]
```diff
@@ -9,16 +9,23 @@ const secretsManagerConfig = localstack ? { endpoint: 'http://localhost:4584', r
 const systemManagerConfig = localstack ? { endpoint: 'http://localhost:4583', region: 'us-west-2' } : {}
 const stsConfig = localstack ? { endpoint: 'http://localhost:4592', region: 'us-west-2' } : {}
 
+const webIdentityCredString = '{credentials: new AWS.TokenFileWebIdentityCredentials()}'
```
Comment (Flydiverny) [inline config/aws-config.js]: Im abit confused about this being a string?

File: config/aws-config.js [hunk 8]
```diff
@@ -9,16 +9,23 @@ const secretsManagerConfig = localstack ? { endpoint: 'http://localhost:4584', r
...
+const irsa = process.env.AWS_IRSA || 0
```
Comment (Flydiverny) [inline config/aws-config.js]:
```suggestion
const irsa = process.env.AWS_IRSA || false
```

**Comment by Flydiverny** [comment]
Looking at https://github.com/aws/aws-sdk-js/issues/2848#issuecomment-540119013, do we still need the `AWS_IRSA` flag and depend on the default credentials chain?

**Comment by Flydiverny** [comment]
Closing this as #200 was merged, please rebase and reopen if you think we need to add the additional things that were in here

**Pull request closed without merging**